### PR TITLE
COR-397: form validation, PART 1, server side

### PR DESCRIPTION
**Pull request by internetti**
Ticket: [COR-397] 

[COR-397]: https://dteam-nrc.atlassian.net/browse/COR-397?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

The backend already sends error messages on form/database/folder creation.
The goal of this PR is showing those errors in the appropriate places at the appropriate times in the pwa.
I also turned some functions returning JSX into components and split the reducer files I touched into multiple files to have a better overview.

The backend also sends errors on record creation, but they're not very useful, so I'll focus on that in the upcoming client side PR.

### CODE REVIEW

File: frontend/apps/pwa/src/features/recorder/RecordEditor.component.tsx [hunk 1]
```diff
@@ -57,6 +66,17 @@ export const RecordEditorComponent: FC<Props> = (props) => {
                     />
                   );
                 })}
+                {_.map(errors, (e: any) => {
```
Comment (internetti) [inline frontend/apps/pwa/src/features/recorder/RecordEditor.component.tsx:69]: this is not final

**Review by internetti — COMMENTED** [review]
I might refactor the FormerContainer and co to also use react-hook-form to show the errors, but this will be very related to the client side validation and to keep this PR "small"(er), I'd do that in the following PRs

File: frontend/apps/pwa/package.json [hunk 2]
```diff
@@ -29,6 +29,7 @@
     "core-api-client": "*",
     "core-auth": "*",
     "date-fns": "^2.25.0",
+    "lodash": "^4.17.21",
```
Comment (neb42) [inline frontend/apps/pwa/package.json:32]: I'm not sure adding lodash is a good idea. It seems to have been used for `map` and `forEach` which have no difference to the native versions. If there was something we wanted lodash for we should install the specific package (eg. `lodash.map`) as lodash is quite big.

File: frontend/apps/pwa/src/features/databases/DatabaseEditor.tsx [hunk 3]
```diff
@@ -11,15 +12,23 @@ type FormData = {
...
 
   const onSubmit = (data: FormData) => {
     client.createDatabase({ object: { name: data.name } }).then((resp) => {
-      if (resp.response) {
+      if (resp.success && resp.response) {
         databaseActions.addOne(resp.response);
         setDatabase(resp.response);
-      } else {
+      } else if (!resp.success && resp.error) {
+        _.forEach(resp.error.details.causes, (e) => {
```
Comment (neb42) [inline frontend/apps/pwa/src/features/databases/DatabaseEditor.tsx:29]: `resp.error.details.causes.forEach((e) => {`

File: frontend/apps/pwa/src/features/former/FieldSections.tsx [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    selectedFieldId,
+    setFieldDescription,
+    setFieldIsKey,
+    setFieldName,
+    setFieldOption,
+    setFieldReferencedDatabaseId,
+    setFieldReferencedFormId,
+    setFieldRequired,
+    setSelectedField,
+  } = props;
```
Comment (neb42) [inline frontend/apps/pwa/src/features/former/FieldSections.tsx:34]:
```
export const FieldSections: React.FC<FieldSectionsProps> = ({
    addField,
    addOption,
    cancelField,
    fields,
    openSubForm,
    removeOption,
    saveField,
    selectedFieldId,
    setFieldDescription,
    setFieldIsKey,
    setFieldName,
    setFieldOption,
    setFieldReferencedDatabaseId,
    setFieldReferencedFormId,
    setFieldRequired,
    setSelectedField,
}) => {
```

File: frontend/apps/pwa/src/features/former/FieldSections.tsx [hunk 5]
```diff
@@ -0,0 +1,91 @@
+import React, { useState } from 'react';
+
+import { FormField } from '../../reducers/Former/types';
+
+import { FieldTypePicker } from './FieldTypePicker';
+import { FormerField } from './FormerField';
+import { FormerProps } from './types';
+
+type FieldSectionsProps = Omit<
```
Comment (neb42) [inline frontend/apps/pwa/src/features/former/FieldSections.tsx:9]: If props are contained to the file I think `Props` suffices as a type name. It's useful to name the props if they're exported and used elsewhere.

File: frontend/apps/pwa/src/features/former/FieldSections.tsx [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+          cancel={() => cancelField(f.id)}
+          errors={f.errors}
+          fieldDescription={f.description}
+          fieldIsKey={f.key}
+          fieldName={f.name}
+          fieldOptions={f.options}
+          fieldRequired={f.required}
+          fieldType={f.fieldType}
+          isSelected={f.id === selectedFieldId}
+          openSubForm={() => openSubForm(f.id)}
```
Comment (neb42) [inline frontend/apps/pwa/src/features/former/FieldSections.tsx:72]: I'd typically define a function:
```
const handleOpenSubForm = (id: string) => () => openSubForm(id);
```
Then:
```
openSubForm={handleOpenSubForm(f.id)}
```

Comment (neb42) [reply frontend/apps/pwa/src/features/former/FieldSections.tsx:72]: As I scrolled down I realised this had been copied from another file, not added new. A bunch of these comments are probably out of scope if things were just moved around.

File: frontend/apps/pwa/src/reducers/Recorder/recorder.adapter.ts [hunk 7]
```diff
@@ -0,0 +1,10 @@
+import { createEntityAdapter } from '@reduxjs/toolkit';
+
+import { FormValue } from './types';
+
+export const adapter = createEntityAdapter<FormValue>({
+  // Assume IDs are stored in a field other than `book.id`
```
Comment (neb42) [inline frontend/apps/pwa/src/reducers/Recorder/recorder.adapter.ts:6]: Are these comments from an example? Should probably make them more relevant or remove them

File: frontend/apps/pwa/src/reducers/Former/former.adapter.ts [hunk 8]
```diff
@@ -0,0 +1,10 @@
+import { createEntityAdapter } from '@reduxjs/toolkit';
+
+import { Form } from './types';
+
+export const adapter = createEntityAdapter<Form>({
+  // Assume IDs are stored in a field other than `book.id`
+  selectId: (form) => form.formId,
+  // Keep the "all IDs" array sorted based on book titles
+  sortComparer: (a, b) => a.formId.localeCompare(b.formId),
```
Comment (neb42) [inline frontend/apps/pwa/src/reducers/Former/former.adapter.ts:9]: Maybe we sort on form name? A uuid isn't a very clear sort order

File: frontend/apps/pwa/src/reducers/Former/former.reducers.ts [hunk 9]
```diff
@@ -0,0 +1,434 @@
...
+import { Form, FormerState, FormField } from './types';
+
+export const reducers = {
+  reset(state: FormerState) {
+    state.entities = {};
+    state.ids = [];
+    const formId = uuidv4();
+    state.selectedFormId = formId;
+    state.selectedFieldId = undefined;
+    adapter.addOne(state, {
```
[the remark this answers is not in the document]
Comment (neb42) [reply frontend/apps/pwa/src/reducers/Former/former.reducers.ts:24]: https://redux-toolkit.js.org/usage/immer-reducers#immutability-and-redux

File: frontend/apps/pwa/src/features/former/FieldSections.tsx (hunk 5 above)
Comment (internetti) [reply frontend/apps/pwa/src/features/former/FieldSections.tsx:9]: I don't think it hurts to give local types a more specific name, but fine

File: frontend/apps/pwa/src/reducers/Recorder/recorder.adapter.ts (hunk 7 above)
Comment (internetti) [reply frontend/apps/pwa/src/reducers/Recorder/recorder.adapter.ts:6]: I didn't add them myself, I'll what I can do ^^

File: frontend/apps/pwa/src/reducers/Former/former.reducers.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (internetti) [reply frontend/apps/pwa/src/reducers/Former/former.reducers.ts:24]: also just copied, but according to this https://redux-toolkit.js.org/usage/immer-reducers#mutating-and-returning-state, it's also valid to mutate the state.
could we do this at another point?

**Pull request merged**